aten/src/ATen/native/cpu/IndexKernel.cpp

```diff
@@ -119,10 +119,158 @@ void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntArrayRef
+    vec.push_back(1);
+  }
+  return vec;
+}
+
+static Tensor restride_dim(
+  const Tensor& src, int64_t dim,
+  IntArrayRef replacement_shape
+) {
+  auto strides = ensure_nonempty_vec(src.strides().vec());
+  strides[dim] = 0;
+  return src.as_strided(replacement_shape, strides);
+}
+
+static TensorIterator make_scatter_iterator(
+  Tensor& self, int64_t dim,
+  const Tensor& index, const Tensor& src
+) {
+  auto index_sizes = ensure_nonempty_vec(index.sizes().vec());
+  auto index_strides = ensure_nonempty_vec(index.strides().vec());
+
+  index_sizes[dim] = 1;
+  index_strides[dim] = 0;
+
+  auto self_restrided = restride_dim(self, dim, index_sizes);
+  auto index_restrided = index.as_strided(index_sizes, index_strides);
+  auto src_restrided = src.as_strided(index_sizes, index_strides);
+
+  auto iter = TensorIterator();
+  iter.dont_compute_common_dtype();
+  iter.dont_resize_outputs();
+  iter.add_output(self_restrided);
+  iter.add_input(src_restrided, src.device(), src.scalar_type());
+  iter.add_input(index_restrided);
+  iter.build();
+  return iter;
+}
+
+template <typename func_t>
+void scatter_base_cpu_kernel(
+  Tensor& self, int64_t dim,
+  const Tensor& index, const Tensor& src,
+  const std::string& method_name,
+  const func_t& f,
+  bool serial_exec = true
+) {
+  // Tests:
+  //  1. index.size(d) <= src.size(d) for all d
+  //  2. index.size(d) <= self.size(d) for all d != dim
+  bool is_wrong_shape = false;
+  int64_t self_dims = ensure_nonempty_dim(self.dim());
+  for (int64_t d = 0; d < self_dims; ++d) {
+    int64_t index_d_size = ensure_nonempty_size(index, d);
+    if (index_d_size > ensure_nonempty_size(src, d)) {
+      is_wrong_shape = true;
+      break;
+    }
+    if (d != dim) {
+      if (index_d_size > ensure_nonempty_size(self, d)) {
+        is_wrong_shape = true;
+        break;
+      }
+    }
+  }
+  TORCH_CHECK(!is_wrong_shape,
+    "Expected index [", index.sizes(), "]"
+    " to be smaller size than src [", src.sizes(), "]"
+    " and to be smaller than self [", self.sizes(), "]"
+    " apart from dimension ", dim
+  );
+
+  DimVector indexed_sizes;
+  DimVector indexed_strides;
+  indexed_sizes.push_back(ensure_nonempty_size(self, dim));
+  indexed_strides.push_back(
+    1//ensure_nonempty_stride(self, dim) * self.element_size()
+  );
+
+  auto iter = make_scatter_iterator(self, dim, index, src);
+
+  auto self_dim_stride = ensure_nonempty_stride(self, dim);
+  auto index_dim_stride = ensure_nonempty_stride(index, dim);
+  auto src_dim_stride = ensure_nonempty_stride(src, dim);
+
+  AT_DISPATCH_ALL_TYPES_AND2(
+    ScalarType::Bool, ScalarType::Half, iter.dtype(),
+    method_name, [&] {
+      cpu_index_kernel<scalar_t>(iter, indexed_sizes, indexed_strides,
+      [&](char* pself, char* psrc, int64_t offset) {
+        auto* self_data = (scalar_t*)pself;
+        const auto* index_data = index.data_ptr<int64_t>()
+          + (psrc - (char*)src.data_ptr()) / sizeof(scalar_t);
```

Comment:
Hm, I'm not sure it works - it implies that offset into index is the same as the offset into src - but is it so if source is larger than index?

Reply:
`src` is restrided in the same way as `index`

Reply:
> Or we just need a new kernel which offsets 3 inputs properly by one step as dictated by their own strides. That is why I do not think we actually need cpu_index_kernel, because looping over dim inside a kernel lambda is faster.
I think this is the best way, you are under no obligation to reuse cpu_index_kernel if it does not provide performance you want. I wonder if there are cases where indexing is much slower than it should be because of the way cpu_index_kernel is structured, but that's a separate issue. I mean, if there is a way to have a single kernel that handles indexing operation and scatter/gather, and provide good performance for all, then that's the best, but if it proves to be too hard, having separate kernels for separate functions is ok (you still can have a single kernel handling scatter/scatter_add/gather and variants I'm pretty sure)

Reply:
Regarding performance of cpu_index_kernel, it seems to index dimensions in the order provided. What about TensorIterator? Does it always loop in a deterministic way? Or it uses some sort of optimizations to decide which dimensions to traversed first?

Reply:
It's slightly complicated, TensorIterator tries to reorder dimensions in such a way that the accesses can be vectorized/coalesced (so e.g. if a and b are contiguous tensors and you are doing a.t() + b.t(), TensorIterator will unscramble a.t() and b.t() again to make operands contiguous). For reductions, reordering is also different. But ignoring this reordering, `iter.for_each` or `iter.serial_for_each` would traverse the tensor with the DimCounter breaking it into 2D slices, so that innermost loop is the last dimension of the tensor. Normally, last dimension is also fastest-varying dimension of the tensor, so accesses have good locality, can be vectorized and hit in cache. But with indexing operations, at least some of the accesses are indirect, and thus all those locality properties are lost.

Reply:
Here's a wiki page with some details on how to use TensorIterator. It's for simpler usecases than yours but can still be useful https://github.com/pytorch/pytorch/wiki/How-to-use-TensorIterator

Reply:
Thank you! Also, cpu_index_kernel is a great example!